dipy/workflows/denoise.py

```diff
@@ -49,7 +69,217 @@ def run(self, input_files, sigma=0, out_dir='',
+            voxels). Default: 2 (denoise in blocks of 5x5x5 voxels).
+        pca_method : string, optional
+            Use either eigenvalue decomposition ('eig') or singular value
+            decomposition ('svd') for principal component analysis. The default
+            method is 'eig' which is faster. However, occasionally 'svd' might
+            be more accurate.
+        tau_factor : float, optional
+            Thresholding of PCA eigenvalues is done by nulling out eigenvalues
+            that are smaller than:
+
+            .. math ::
+
+                    \tau = (\tau_{factor} \sigma)^2
+
+            \tau_{factor} can be change to adjust the relationship between the
+            noise standard deviation and the threshold \tau. If \tau_{factor}
+            is set to None, it will be automatically calculated using the
+            Marcenko-Pastur distribution [2]_.
+            Default: 2.3 (according to [1]_)
+        out_dir : string, optional
+            Output directory (default input file directory)
+        out_denoised : string, optional
+            Name of the resulting denoised volume (default: dwi_lpca.nii.gz)
+
+        References
+        ----------
+        .. [1] Veraart J, Novikov DS, Christiaens D, Ades-aron B, Sijbers,
+        Fieremans E, 2016. Denoising of Diffusion MRI using random
+        matrix theory. Neuroimage 142:394-406.
+        doi: 10.1016/j.neuroimage.2016.08.016
+
+        .. [2] Veraart J, Fieremans E, Novikov DS. 2016. Diffusion MRI noise
+        mapping using random matrix theory. Magnetic Resonance in
+        Medicine.
+        doi: 10.1002/mrm.26059.
+
+        .. [3] Manjon JV, Coupe P, Concha L, Buades A, Collins DL (2013)
+        Diffusion Weighted Image Denoising Using Overcomplete Local
+        PCA. PLoS ONE 8(9): e73021.
+        https://doi.org/10.1371/journal.pone.0073021
+
+        """
+        io_it = self.get_io_iterator()
+        for dwi, bval, bvec, odenoised in io_it:
+            logging.info('Denoising {0}'.format(dwi))
+            data, affine, image = load_nifti(dwi, return_img=True)
+
+            if not sigma:
+                logging.info('Estimating sigma')
+                bvals, bvecs = read_bvals_bvecs(bval, bvec)
+                gtab = gradient_table(bvals, bvecs, b0_threshold=b0_threshold,
+                                      atol=bvecs_tol)
+                sigma = pca_noise_estimate(data, gtab, correct_bias=True,
+                                           smooth=3)
+                logging.debug('Found sigma {0}'.format(sigma))
+
+            denoised_data = localpca(data, sigma=sigma,
+                                     patch_radius=patch_radius,
+                                     pca_method=pca_method,
+                                     tau_factor=tau_factor)
+            save_nifti(odenoised, denoised_data, affine, image.header)
+
+            logging.info('Denoised volume saved as {0}'.format(odenoised))
```

Comment:
you may use `logging.info('Denoised volume saved as %s', odenoised)`
see: https://reinout.vanrees.org/weblog/2015/06/05/logging-formatting.html
"Don’t use the `'{}'.format()`, as it formats the string before it gets passed to the logger. So even if the log message doesn’t need to be actually logged somewhere, the full string gets created.
The proposed example only gets passed a string and some extra arguments and only turns it into a real full string for in your logfile if it actually gets logged. So if you only display WARN level and higher, your DEBUG messages don’t need to be calculated."

Reply:
but there is no time difference
```
$ python -mtimeit -s"import logging; logging.info('sample %s', 'text')"
100000000 loops, best of 3: 0.0121 usec per loop
$ python -mtimeit -s"import logging; logging.info('sample {}'.format('text'))"
100000000 loops, best of 3: 0.0127 usec per loop
```

Reply:
Thank you for this information. I believe that `%` was the **old style** and we should avoid it. I do not know what is the official rule and your reference start to be quite old (2015). I will have a look atthis before updating this code

Reply:
I would say that in case you have lots of logging and not all are always needed - levels info/debug, then the formating output at the time it is really needed (just-in-time) can be useful...

Reply:
After a quick check on the Python doc, your reference is still up to date. They keep this behavior for backward compatibility and it seems they will keep it for a long time.... there is a workaround but I prefer to update the syntax. Thanks @borda, good to know